drivers/input/input_adc_keys.c

```diff
@@ -0,0 +1,228 @@
+	}
+}
+
+static void adc_keys_work_handler(struct k_work *work)
+{
+	struct k_work_delayable *dwork = CONTAINER_OF(work, struct k_work_delayable, work);
+	struct adc_keys_data *data = CONTAINER_OF(dwork, struct adc_keys_data, dwork);
+	const struct device *dev = data->self;
+	const struct adc_keys_config *cfg = dev->config;
+
+	adc_keys_process(dev);
+
+	k_work_schedule(&data->dwork, K_MSEC(cfg->sample_period_ms));
+}
+
+static int adc_keys_init(const struct device *dev)
+{
+	const struct adc_keys_config *cfg = dev->config;
+	struct adc_keys_data *data = dev->data;
+	int ret;
+
+	if (!device_is_ready(cfg->channel.dev)) {
+		LOG_ERR("ADC controller device %s not ready", cfg->channel.dev->name);
+		return -ENODEV;
+	}
+
+	ret = adc_channel_setup_dt(&cfg->channel);
+	if (ret) {
+		LOG_ERR("ADC channel setup failed %d", ret);
+		return ret;
+	}
+
+	adc_sequence_init_dt(&cfg->channel, &data->seq);
+
+	data->self = dev;
+	k_work_init_delayable(&data->dwork, adc_keys_work_handler);
+
+#if defined(CONFIG_INPUT_LOG_LEVEL_DBG)
+	for (uint8_t i = 0; i < cfg->code_cnt; i++) {
+		LOG_DBG("* code %d: key_index=%d threshold=%d mV code=%d", i,
+			cfg->code_cfg[i].key_index, cfg->code_cfg[i].press_mv,
+			cfg->key_code[cfg->code_cfg[i].key_index]);
+	}
+#endif /* CONFIG_INPUT_LOG_LEVEL_DBG */
+
+	k_work_schedule(&data->dwork, K_MSEC(cfg->sample_period_ms));
+
+	return 0;
+}
+
+#define ADC_KEYS_CODE_CFG_ITEM(node_id, prop, idx)                                                 \
+	{                                                                                          \
+		.key_index = DT_NODE_CHILD_IDX(node_id) /* include disabled nodes */,              \
+		.press_mv = DT_PROP_BY_IDX(node_id, prop, idx),                                    \
+	}
+
+#define ADC_KEYS_CODE_CFG(node_id)                                                                 \
+	DT_FOREACH_PROP_ELEM_SEP(node_id, press_thresholds_mv, ADC_KEYS_CODE_CFG_ITEM, (,))
+
+#define ADC_KEYS_KEY_CODE(node_id) DT_PROP(node_id, zephyr_code)
+
+#define ADC_KEYS_INST(n)                                                                           \
+	static struct adc_keys_data adc_keys_data_##n;                                             \
+                                                                                                   \
+	static const struct adc_keys_code_config adc_keys_code_cfg_##n[] = {                       \
+		DT_INST_FOREACH_CHILD_STATUS_OKAY_SEP(n, ADC_KEYS_CODE_CFG, (,))};                 \
+                                                                                                   \
+	static const uint16_t adc_keys_key_code_##n[] = {                                          \
+		DT_INST_FOREACH_CHILD_SEP(n, ADC_KEYS_KEY_CODE, (,))};                             \
+                                                                                                   \
+	static struct adc_keys_key_state                                                           \
```

Comment:
Can we constify this structure as well?

Reply:
No. It stores the state of each key and will be compared in each iteration.